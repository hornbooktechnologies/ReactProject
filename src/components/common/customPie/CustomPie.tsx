//@ts-nocheck
import * as d3 from "d3";
import { useEffect, useRef } from "react";
import { ChartColors } from "../../../utils/AppConstants";
import { getChartFillValue, getRatioChangePct, isPositiveRatio } from "../../../utils/DonutChartUtils";
import RelativeRatio from "./RelativeRatio";
import PieRatioValue from "./PieRatioValue";

type Props = {
    chartGradientType?: string
    gradientColor1?: string
    gradientColor2?: string
    type?: string
    height?: number
    width?: number
    color?: string
    data: ChartData
}

type ChartData = {
    series1?: any
    series2?: any
    value?: number
    plan?: number
    forecast?: number
}

const CustomPie = (props: Props) => {
    const chartContainerRef = useRef();
    const { data, width, height,chartType } = props;
    const { chartGradientType, gradientColor1, gradientColor2, color, plan = 0, forecast = 0, value, series1, series2,middleRatio } = data;
    const ptPosCoef = 0.1;
    const chartFillValue = getChartFillValue(value, plan, forecast);
    const changePct = getRatioChangePct(plan, forecast);
    const showGreenRatio = isPositiveRatio(chartType, plan, forecast);
    useEffect(() => {
      const gradientId = `pieGradient-${chartGradientType}`;
      const drawChart = () => {
        // Remove the previous svg
        d3.select(chartContainerRef.current).select("svg").remove();
        // Create new svg
        const svgElement = d3
          .select(chartContainerRef.current)
          .append("svg")
          .attr("width", width)
          .attr("height", height)
          .append("g")
          .attr("transform", `translate(${width / 2}, ${height / 2})`);
  
        const gradientElement = svgElement.append("defs")
          .append("linearGradient")
          .attr("id", gradientId);
       
        gradientElement.append("stop")
          .attr("offset", "0%")
          .attr("stop-color", gradientColor1);
  
        gradientElement.append("stop")
          .attr("offset", "90%")
          .attr("stop-color", gradientColor2);
        
        // this is the width of the pie chart arc
        const arcWidth = 10;
        const boxSize = Math.min(width, height);
        const radius = boxSize / 2;
        const bgRad = arcWidth / 2;
        const bgArc = d3
          .arc()
          .outerRadius(radius - (bgRad - 0.5))
          .innerRadius(radius - (bgRad + 0.5));
  
        const fgArc = d3
          .arc()
          .outerRadius(radius)
          .innerRadius(radius - arcWidth)
          .cornerRadius(arcWidth);
  
        const pie = d3
          .pie()
          .sort(null)
          .value((d) => d.value);
  
        const bgData = [{ color: `${ChartColors._ccc}`, value: 100 }];
        svgElement
          .selectAll(".background")
          .data(pie(bgData))
          .enter()
          .append("path")
          .attr("class", "background")
          .style("fill", (d) => d.data.color)
          .attr("stroke", `${ChartColors._455054}`)
          .style("stroke-width", 8)
          .attr("d", bgArc);
  
        const fgData = [];
        if (chartFillValue >= 0) {
          fgData.push({ color: `url(#${gradientId})`, value: chartFillValue })
          fgData.push({ color: 'transparent', value: 100 - chartFillValue })
        } else {
          fgData.push({ color: 'transparent', value: 100 + chartFillValue })
          fgData.push({ color: `url(#${gradientId})`, value: - chartFillValue })
        }
  
        const fgPie = svgElement
          .selectAll(".foreground")
          .data(pie(fgData))
          .enter()
          .append("path")
          .attr("class", "foreground")
          .style("stroke-width", 8)
          .style("fill", (d) => d.data.color);
        
        const ptRad = radius - arcWidth / 2;
        let pt = null;
        let arro = null;
  
        const hp = Math.PI / 2;
        fgPie
          .transition()
          .duration(1000)
          .attrTween("d", (d) => {
            const fromAngle = chartFillValue >= 0 ? 0 : 2 * Math.PI;
            const ip = d3.interpolate(
              { startAngle: fromAngle, endAngle: fromAngle },
              d
            );
            return (t) => {
              const arcData = ip(t);
              if (pt && d.data.color !== "transparent") {
                let rad = 0;
                if (chartFillValue > 0) {
                  let cr = hp + arcData.endAngle / 2;
                  if (cr > hp + ptPosCoef) {
                    cr = arcData.endAngle >= 2 * Math.PI ? hp : hp + ptPosCoef;
                  }
                  rad = arcData.endAngle - cr;
                } else {
                  let cr = hp - (2 * Math.PI - arcData.startAngle) / 2;
                  if (cr < hp - ptPosCoef) {
                    cr = arcData.startAngle <= 0 ? hp : hp - ptPosCoef;
                  }
                  rad = arcData.startAngle - cr;
                }
                pt.attr("cx", ptRad * Math.cos(rad));
                pt.attr("cy", ptRad * Math.sin(rad));
                if (chartFillValue > 0) {
                  arro.attr(
                    "transform",
                    `translate(${ptRad * Math.cos(rad)},${
                      ptRad * Math.sin(rad)
                    }) rotate(${chartFillValue * 3.6}) `
                  );
                } else {
                  arro.attr(
                    "transform",
                    `translate(${ptRad * Math.cos(rad)},${
                      ptRad * Math.sin(rad)
                    }) rotate(${chartFillValue * 3.6 + 180}) `
                  );
                }
              }
              return fgArc(arcData);
            };
          });
      };
      drawChart();
    }, [chartContainerRef, color, chartFillValue, height, width,chartGradientType,gradientColor1,gradientColor2]);

    return (
      <>
       <div className="uk-flex uk-flex-center">
        <div
          className="chart-continer"
          style={{ width: width, height: height }}
        >
          <div className="dount-chart-container">
            <div ref={chartContainerRef} />
          </div>
          <div className={`${props.customClassForRatio ? props.customClassForRatio:''} donutmiddleData`}>
            {middleRatio ? (
              <PieRatioValue {...data} />
            ) : (
              <RelativeRatio
                changePercent={changePct}
                isPositiveRatio={showGreenRatio}
                equationX={series1.title}
                equationY={series2.title}
              />
            )}
          </div>
        </div>
        </div>
      </>
    )
}

export default CustomPie