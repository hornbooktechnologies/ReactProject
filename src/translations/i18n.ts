import { initReactI18next } from 'react-i18next'
import i18n from 'i18next'
import LanguageDetector from 'i18next-browser-languagedetector'
import { TRANSLATIONS_EN } from './en/Translations'
import { TRANSLATIONS_JA } from './ja/Translations'

i18n
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    lng: 'en',
    resources: {
      en: {
        translation: TRANSLATIONS_EN,
      },
      ja: {
        translation: TRANSLATIONS_JA,
      },
    },
  })

export default i18n
