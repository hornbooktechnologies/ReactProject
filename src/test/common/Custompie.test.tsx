import React from "react";
import CustomPie from "../../components/common/customPie/CustomPie";
import renderer from "react-test-renderer";
import i18n from "../../i18nForTests";
import { I18nextProvider } from "react-i18next";
describe("Component: CustomPie", () => {
  var ChartData = {
    series1: {
      title: "plan",
      seriesValue: 50,
      prefix: "",
      suffix1: "",
      suffix2: "",
    },
    series2: {
      title: "plan",
      seriesValue: 50,
      prefix: "",
      suffix1: "",
      suffix2: "",
    },
    value: 20,
    plan: 20,
    forecast: 10,
  };

  var Props = {
    chartGradientType: "green",
    gradientColor1: "#000",
    gradientColor2: "#fff",
    type: "large",
    height: 190,
    width: 190,
    color: "#ddd",
    data: ChartData,
  };
  

  it("CustomPie renders correctly", () => {
    const tree = renderer
      .create(
        <I18nextProvider i18n={i18n}>
          <CustomPie {...Props} />
        </I18nextProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
