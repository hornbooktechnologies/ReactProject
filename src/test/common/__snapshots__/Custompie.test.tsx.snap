// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CustomPie CustomPie renders correctly 1`] = `
<div
  className="uk-flex uk-flex-center"
>
  <div
    className="chart-continer"
    style={
      Object {
        "height": 190,
        "width": 190,
      }
    }
  >
    <div
      className="dount-chart-container"
    >
      <div />
    </div>
    <div
      className=" donutmiddleData"
    >
      <div
        className="uk-text-center thumb-font"
      >
        <div
          className="jera-dislike-font"
        >
          <i
            className="jera color-red-60 jera-dislike"
          />
        </div>
        <div
          className="font-prvalue color-red-60"
        >
          <span>
            50
          </span>
          <span
            className="font-12 vertical-top"
          >
            %
          </span>
        </div>
        <div>
          <span
            className="font-12 totalRatio"
          >
            CHARTS.DONUT.TOOLTIP_HEAD
          </span>
          <span
            className="uk-tooltip-cutom tooltip-text"
          >
            <i
              className="jera-info-circle icn-16 color-blue-70 top-2 ml-3"
            />
            <div
              className="uk-tooltip uk-animation-scale-up uk-animation-center uk-transdorm-origin-center-left uk-togglabe-enter uk-tooltip-bottom uk-tooltip-bottom-center uk-transform-origin-center-left uk-text-left font-10"
            >
              <span
                className="color-grey-blue-80 uk-display-block"
              >
                <strong
                  className="color-grey-blue-95"
                >
                  CHARTS.DONUT.TOOLTIP_MAIN_TEXT
                </strong>
                Â  
                CHARTS.DONUT.TOOLTIP_SUB_TEXT
              </span>
              <span
                className="color-red-60 mt-15 uk-display-block"
              >
                CHARTS.DONUT.EQUATION
              </span>
              <span
                className="color-grey-blue-80 uk-display-block"
              >
                CHARTS.DONUT.EQUATION_LHS
                 = Y (
                plan
                ) / X (
                plan
                )
              </span>
            </div>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
